osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs

```diff
@@ -55,30 +72,249 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+                                     * (1 + 0.1 * Math.Min(1, (attributes.NoteCount + attributes.HoldNoteCount) / 1500.0)); // Length bonus, capped at 1500 notes
 
             return difficultyValue;
         }
 
-        private double totalHits => countPerfect + countOk + countGreat + countGood + countMeh + countMiss;
+        private double totalJudgements => countPerfect + countOk + countGreat + countGood + countMeh + countMiss;
+        private double totalSuccessfulJudgements => countPerfect + countOk + countGreat + countGood + countMeh;
 
-        /// <summary>
-        /// Accuracy used to weight judgements independently from the score's actual accuracy.
-        /// </summary>
         private double calculateCustomAccuracy()
         {
-            if (totalHits == 0)
+            if (totalJudgements == 0)
                 return 0;
 
-            return (countPerfect * 320 + countGreat * 300 + countGood * 200 + countOk * 100 + countMeh * 50) / (totalHits * 320);
+            return (countPerfect * 320 + countGreat * 300 + countGood * 200 + countOk * 100 + countMeh * 50) / (totalJudgements * 320);
+        }
+
+        /// <summary>
+        /// Returns the estimated tapping deviation of the score, assuming the average hit location is in the center of the hit window.
+        /// </summary>
+        private double? computeEstimatedUr(ManiaDifficultyAttributes attributes)
+        {
+            if (totalSuccessfulJudgements == 0 || attributes.NoteCount + attributes.HoldNoteCount == 0)
+                return null;
+
+            // Lazer LN heads are the same as Notes, so return NoteCount + HoldNoteCount for lazer scores.
+            double logNoteCount = isLegacyScore ? Math.Log(attributes.NoteCount) : Math.Log(attributes.NoteCount + attributes.HoldNoteCount);
+            double logHoldCount = Math.Log(attributes.HoldNoteCount);
+
+            double likelihoodGradient(double d)
+            {
+                if (d <= 0)
+                    return 0;
+
+                JudgementProbs pNotes = pNote(d);
+                // Since lazer tails have the same hit behaviour as Notes, return pNote instead of pHold for them.
+                JudgementProbs pHolds = isLegacyScore ? pHold(d) : pNote(d, tail_multiplier);
+
+                return -totalProb(pNotes, pHolds, logNoteCount, logHoldCount);
+            }
+
+            // Finding the minimum of the function returns the most likely deviation for the hit results. UR is deviation * 10.
+            double deviation = FindMinimum.OfScalarFunction(likelihoodGradient, 30);
+
+            return deviation * 10;
+        }
+
+        private double[] getLegacyHitWindows(ScoreInfo score, ManiaDifficultyAttributes attributes)
+        {
+            double[] legacyHitWindows = new double[5];
+
+            double overallDifficulty = attributes.OverallDifficulty;
+
+            if (attributes.Convert)
+                overallDifficulty = 10;
```

Comment:
according to the information that will be added in https://github.com/ppy/osu-wiki/pull/8245, this can actually be either OD 10 or OD ~5 (see [wiki/Gameplay/Judgement/osu!mania/en.md](https://github.com/ppy/osu-wiki/commit/3795d6d161a18dde46cc2c1a30dd6835e71b1a0f#diff-182fdf1bbbc79c215c413a619f2aacd28cb4f82e32cfa813c5e3d0eb7d101171)). this should definitely be accounted for.

aside from that, this is definitely going to need a comment. i've discussed this with natelytle in pp dev discord and here's a potential comment i can propose:

```
When converting beatmaps to osu!mania in stable, the resulting hit window sizes are dependent on whether the beatmap's OD is above or below 4.
```

Reply:
Done.